[28기 김정현][ADD] feed 영역 상단 friends 영역 스크롤 이벤트 추가

KimJeongHyun opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 단기적 목표 : main.html navigation bar buttons modal 구현
- 장기적 목표 : display:flex를 통한 레이아웃 나누기 익숙해지기, JS 동작을 마치 캡슐화된 형태로 다루기
<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
0. Mission 달성 확인
- [x] Mission1 : 로그인 페이지 레이아웃
- [x] Mission2 : ID, PW 입력 시 로그인 버튼 활성화 기능
- [x] Mission3 : 메인 페이지 레이아웃
- [x] Mission4 : 댓글 내용 입력 후 Enter press, 혹은 게시 버튼 클릭시 댓글 추가 기능
- [x] Mission5 : ID, PW Validation
- [x] Mission6 : 댓글 좋아요 / 삭제 기능
- [x] Mission7 : 아이디 검색 기능
- [x] Mission8 : nav 프로필 사진 클릭 시 메뉴 박스 생성
- [x] Mission9 : 반응형 구현

1. 로그인 입력 데이터 유효성 검사 
- 입력창에 변화가 일어날 때 마다 유효성 검사 함수가 실행된다.
- 유효성 검사 함수의 조건문에 따라 버튼의 배경색과 disabled 속성이 토글된다.
- 테스트 방법 : `/index` 페이지 이동 >>> 로그인 input 데이터 입력

2. 브라우저 resize에 따른 index.html element 변화
 - window.innerWidth<=735px 인 경우 좌측 imgBox를 숨김
 - window.innerWidth>735px이면 좌측 imgBox가 표시됨
 - 테스트 방법 : `/index` 페이지 이동 >>> 브라우저 크기 조절

3. 브라우저 resize에 따른 main.html element 변화
 - window.innerWidth<=627 인 경우 검색 바를 숨김
 - window.innerWidth>627 인 경우 검색 바가 표시됨
 - window.innerWidth>690 && window.innerWidth<=1001인 경우 side 네비게이션 바를 숨김. margin-left를 vw값으로 할당하여 부드러운 resizing 처리
 - 
 - window.innerWidth<690인 경우 margin-left를 0으로 할당해 브라우저 크기에 맞게 들어오게 배치
 - 이외의 경우 margin-left 및 display를 초기화 ( margin-left:0px, display:block )
 - 테스트 방법 : `/main` 페이지 이동 >>> 브라우저 크기 조절

4. 검색 바 애니메이션
 - 검색 바를 focusin하면 돋보기 아이콘이 사라지며 부드럽게 커서가 좌측으로 이동함
 - 검색 바가 focusout되면 돋보기 아이콘이 표시되며 부드럽게 커서가 우측으로 이동함
 - 테스트 방법 : `/main` 페이지 이동 >>> 검색 바 클릭 후 다른 영역 클릭

5. 피드 내용 확장하기 (더 보기)
 - 최초 로드시 각 피드의 내용들을 tempArray에 push
 - 최초 로드시 피드 중 10글자가 넘는 내용이 있는 경우, slice(0,10).trim()처리를 통해 글자 수를 줄여서 표시함
 - '더 보기' 버튼의 onclick으로 feedExtendContents 함수 호출
 - 호출된 함수는 해당 피드의 전체 내용을 tempArray를 통해 로드함
 - 테스트 방법 : `/main` 페이지 이동 >>> 각 피드 '더 보기' 버튼 클릭
 
6. 댓글 게시 버튼 활성화
 -  input이 발생할 때마다 각 피드의 댓글 input에 공백이 아닌 글자가 있는지 검사함.
 - 공백이 없는 경우 버튼의 투명도를 1로 조정
 - 공백만 있는 경우 버튼의 투명도를 0.3으로 조정
 - 테스트 방법 : '/main' 페이지 이동 >>> 댓글 input 데이터 입력

7. 댓글 달기
 - 게시 버튼을 누르거나, 댓글 textarea에 focus된 상태에서 엔터를 누르는 경우 addFeedComment(event) 함수가 호출됨
 - addFeedComment(event)는 event가 발생한 태그의 name을 통해 몇 번째 피드의 댓글인지 확인함
 - 해당 피드의 댓글 textarea의 value를 댓글 박스에 appendChild()를 통해 추가함.
 - 테스트 방법 : '/main' 페이지 이동 >>> 댓글 textarea 데이터 입력 후 게시 버튼 클릭 또는 Enter 

8. 우측 사이드 바 글자 축소
 - '회원님을 팔로우합니다.' 또는 'xxx님 외 n명이 팔로우합니다'에 해당하는 innerHTML을 검사함
 - 만약 24를 초과하는 length가 감지되는 경우 slice(0,25).trim()을 적용함.

9. ID 검색 기능
 - 상단 검색 바에 1 이상 길이의 텍스트 입력시, ID 배열에서 해당 텍스트가 include된 요소를 확인함
 - 확인된 요소가 있는 경우, 검색 플로팅 영역에 표시함
 - backspace 입력시 플로팅 영역의 요소에 대해 include 검사 진행
 - false인 경우 해당 요소를 삭제함
 - 테스트 방법 : '/main' 페이지 이동 >>> 검색 input 데이터 입력 또는 backspace

10. 메인 화면 이동
 - 상단 로고 및 홈 버튼을 통해 /main 페이지로 이동 가능
 - 테스트 방법 : '/main' 페이지 이동 >>> 상단 로고 또는 홈 버튼 클릭
<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
 - display:flex의 적절한 활용을 통해 마치 각 영역이 독립된 필드인 것처럼 취급하는 것에 영감을 얻었습니다.
 - ID 입력시 '@'뿐만 아니라 .com, .net 등의 접미어를 어떻게 똑똑하게 검사할지 고민하고 있습니다.
 - semantic tag를 적절히 이용해 HTML 코드의 가독성이 전보다 향상된 것 같습니다.
 - 반응형에 대해 고민이 많았는데, 이번 레이아웃 클로닝을 통해 어느정도 해결되었습니다.

<br />

## :: 기타 질문 및 특이 사항
 - 전역 변수를 할당하는 것은 위험하다는 것을 알고 있으나 예를 들어 함수가 호출될 때마다 document.getElement~ 를 통해 DOM 트리 순회가 계속 발생된다면 비효율적이지 않을까? 하는 생각이 있습니다. 
 - 실제 인스타그램은 스크롤 이벤트에 따라 피드를 로드하여 HTML에 그려주는 것으로 보이는데, 계속 스크롤을 내리다보면 크롬 브라우저가 점점 무거워집니다. 이 부분에 대해서 현재 브라우저의 스크롤 기준으로 피드를 위 아래로 2~3개 정도만 노출시키고 다른 피드는 숨기는 방법에 대해 고민 중입니다.

bonbon0202 commented:
주민님! 템플릿 체크리스트 작성해주시면 리뷰해드리도록 하겠습니다 :)

rhdwnals1 commented:
> 주민님! 템플릿 체크리스트 작성해주시면 리뷰해드리도록 하겠습니다 :)

체크했습니다! 감사합니다ㅎㅎㅎㅎ

bonbon0202 commented on students/14th/juminkong/common.css:
```diff
@@ -0,0 +1,14 @@
+* {
+    box-sizing: border-box;
+    padding: 0;
+    margin: 0;
+}
```
common.css 잘 활용해 주셨네요!! 위 내용은 common.css 보다는 브라우저의 style 기본값을 초기화시키기 위해 작성하는 코드를 담는 reset.css 파일에 작성해주시면 더 적합할 것 같습니다 :)

rhdwnals1 replied:
넵 reset.css에 작성하겠습니다!

bonbon0202 commented on students/14th/juminkong/login.css:
```diff
@@ -0,0 +1,73 @@
+
```
불필요한 빈 줄은 삭제해주세요!!

rhdwnals1 replied:
넵!

bonbon0202 commented on students/14th/juminkong/login.css:
```diff
@@ -0,0 +1,73 @@
+
+.container {
+    width: 350px;
+    height: 380px;
+    margin: 100px auto;
+    display: flex;
+    flex-direction: column;
+    align-content: center;
+    flex-wrap: wrap;
+    background-color: #fff;
+    border: 1px solid #DBDBDB;
```
와우!! css 속성 반영 잘 해주셨네요!! ㅎㅎㅎ 굿굿! 👍🏻
border 속성도 레이아웃을 만들어주는 width, height, margin 다음에 함께 작성해주시면 더 좋을 것 같습니다 :)

rhdwnals1 replied:
네 수정했습니다!

bonbon0202 commented on students/14th/juminkong/login.css:
```diff
@@ -0,0 +1,73 @@
+}
+
+img{
+    align-self: center;
+    width: 175px;
+    margin: 35px 0;
```
위 속성도 css 작성 순서를 지켜서 작성해주세요 :) 
1. 레이아웃이 형성되는데 영향을 많이 미치는 순서 
2. 연관된 순서는 인접하게 작성

rhdwnals1 replied:
감사합니다아 👍

bonbon0202 commented on students/14th/juminkong/login.css:
```diff
@@ -0,0 +1,73 @@
+    color: #8e8e8e;    
+}
+
+input:focus {
+    border: 1px solid #8a8a8a;
+    outline: none;
+}
```
focus 가상선택자 활용 굿굿!

bonbon0202 commented on students/14th/juminkong/login.css:
```diff
@@ -0,0 +1,73 @@
+    outline: none;
+    opacity: 0.5;
+}
+
```
선택자 간 사이는 한 줄 만 띄어주세요!:)

bonbon0202 commented on students/14th/juminkong/login.css:
```diff
@@ -0,0 +1,73 @@
+a {
+    text-decoration: none;
+}
+
```
여기도 마찬가지 입니다!!

bonbon0202 commented on students/14th/juminkong/login.html:
```diff
@@ -0,0 +1,24 @@
+    <link href="login.css" rel="stylesheet" type="text/css" />
+    <link href="common.css" rel="stylesheet" type="text/css" />
+  </head>
+
+
```
불필요한 빈 줄은 삭제해주세요!!

bonbon0202 commented on students/14th/juminkong/login.html:
```diff
@@ -0,0 +1,24 @@
+
+
+  <body>
+        <div class="container">
+        <img src= "logo_text.png" alt="instagram_logo">
+        <input type="text" class="input_ID" placeholder="전화번호, 사용자 이름 또는 이메일">
+        <input type="password" class="input_PW" placeholder="비밀번호">
+        <button type="button" class="btn">
+        <span>로그인</span>
+        </button>
+        <span class="forget"><a href="https://www.instagram.com/accounts/password/reset/">비밀번호를 잊으셨나요?</a></span>
+        </div>
```
들여쓰기 지켜주세요!! 가독성을 위해 들여쓰기 매우 중요합니다 :) 
더불어 들여쓰기는 4tab으로 하면 코드가 너무 가로로 길어지기 때문에 2tab이 컨벤션입니다. 

```suggestion
        <div class="container">
          <img src= "logo_text.png" alt="instagram_logo">
          <input type="text" class="input_ID" placeholder="전화번호, 사용자 이름 또는 이메일">
          <input type="password" class="input_PW" placeholder="비밀번호">
          <button type="button" class="btn">
            <span>로그인</span>
          </button>
          <span class="forget">
            <a href="https://www.instagram.com/accounts/password/reset/">비밀번호를 잊으셨나요?</a>
          </span>
        </div>
```

bonbon0202 commented on students/14th/juminkong/login.html:
```diff
@@ -0,0 +1,24 @@
+        </button>
+        <span class="forget"><a href="https://www.instagram.com/accounts/password/reset/">비밀번호를 잊으셨나요?</a></span>
+        </div>
+    <script src="script.js"></script>
```
script태그 아래에 잘 작성해주셨네요 :) !!

bonbon0202 commented on students/14th/juminkong/script.js:
```diff
@@ -0,0 +1,12 @@
+let input_ID = document.querySelector('.input_ID');
+let input_PW = document.querySelector('.input_PW');
+
+input_ID.addEventListener('keydown', () => {
+    let button = document.querySelector('button');
+    button.style.opacity = 1;
+});
+
+input_PW.addEventListener('keydown', () => {
+    let button = document.querySelector('button');
+    button.style.opacity = 1;
+});
```
하나의 코드로 합쳐 볼 수 있을 것 같아요!! :)

bonbon0202 reviewed (commented):
주민님의 첫 pr을 축하드립니다!! 🥳
전체적으로 코드 깔끔하게 잘 작성해주셨네요!! :) 
세션 때 알려드린 내용을 열심히 적용해보셨다는게 느껴집니다! 
리뷰 반영해주시고 메인 레이아웃과 기능구현도 함께 해보세요!! :) 고생하셨습니다

KimJeongHyun committed:
[Add] Login, Main 페이지 레이아웃
 - Login, Main 페이지 html 작성 완료
 - 유효성 검사 로직 작성 완료
 - resize event 로직 작성 완료
 - Main 페이지 댓글 달기 구현 완료

KimJeongHyun committed:
[Fix] 로그인 버튼 유효성 검사 - disabled 속성 토글 추가

KimJeongHyun committed:
[Fix] 로그인 버튼 유효성 검사 - disabled 토글 추가 완성

KimJeongHyun committed:
[Fix] index.html footer 의미 없는 글자 삭제

KimJeongHyun committed:
[Fix] mainJS.js 의미없는 display:none 처리 제거

KimJeongHyun committed:
[Update] main 페이지 검색 바 로직 작성.
 - 검색 바의 value에 따라 ID 배열에서 includes되는 것을 찾아 표시합니다.
 - input 및 keyup===Backspace일 때 현재 value가 includes되는지 아닌지
   검사하여 표시 여부를 결정합니다.
 - 중첩 append를 막기 위해 input시 영역에 중복되는 value가 있는지
   검사하고, 있다면 append하지 않으며 없다면 append합니다.
 - Backspace시 현재 영역에 있는 node 중 includes(value)가 false인 값이
   있다면 해당 node를 제거합니다.
 - Backspace시 value.length === 0일 때는 영역의 모든 노드를 지웁니다.

KimJeongHyun committed:
[Fix] 코드 단락 수정

KimJeongHyun committed:
[Fix] 코드 단락 수정

KimJeongHyun committed:
[FIX] 코드 단락 수정

KimJeongHyun committed:
Merge branch 'feature/jeonghyunkim' of https://github.com/wecode-bootcamp-korea/westagram-frontend into feature/jeonghyunkim

KimJeongHyun committed:
[ADD] main.js navBtn 로직 추가 및 댓글 달기 로직 수정
 - navBtn 클릭 시 전체를 가리는 hideContainer 표시됨
 - hideContainer click시 hide 및 floatingBtnArea가 제거됨
 - 댓글 Enter 및 버튼 클릭시 '게시'버튼 활성화/비활성화 적용
 - 댓글 달기 함수 호출시 textarea의 개행 문자 제거

KimJeongHyun committed:
[FIX] HTML, JS, CSS Refactoring
 - bottom->up 구조로 개편에 따라 불필요한 CSS 요소 및 JS style 변경 제거
 - navBtn arrow의 left 속성 변경방법성 transform:translateX로 대체
 - HTML 단락 정리
 - CSS 단락 정리

KimJeongHyun committed:
[FIX] JS, CSS Refactoring
 - landingJS.js -> indexJS.js 로 이름 변경
 - mainJS.js 코드 작성 순서 변경 ( window -> nav -> feed -> side )
 - JS 적용 영역별 주석 추가
 - main.html 검색 결과 영역 arrow 요소 배치 변경

KimJeongHyun committed:
[FIX] mainStyle.css Refactoring
 - 불필요한 flex 속성 제거
 - z-index 순서 정리
   :: 버튼 플로팅 발생시 navBtn 제외 nav 요소를 접근할 수 있도록 정리함

KimJeongHyun committed:
[ADD] feed 상단 friends 영역 스크롤 이벤트 추가
 - friendsFeed clientWidth, friendsFeedUl offsetWidth 차이 이용
 - Ul이 부모 요소보다 길면 오른쪽 버튼 자동으로 표시. 그렇지 않으면
   default hidden.
 - scrollLeft===0일시 오른쪽 버튼 표시
 - scrollLeft===clientWidth-offsetWidth이면 왼쪽 버튼 표시
 - scrollLeft가 0<~<끝점 범위에 있는 경우 양쪽 버튼 모두 표시
